### PR TITLE
resource_auditor: add audit for HEAD default branch

### PR DESCRIPTION
**Pull request by nandahkrishna**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

This PR adds an audit to ensure that the correct default branch for a Git repository is specified for `head` when using `--strict --online`. For `main` and other alternatives, we need to explicitly specify the default branch using `branch:`, as the default is `master`.

Adding an audit will ensure this is checked in PRs to add new formulae.

I haven't written a test for this audit as it would require network access (and use a real repository URL), but I'd be happy to add one if that's preferable.

(CC @vladimyr)

### CODE REVIEW

File: Library/Homebrew/resource_auditor.rb [hunk 1]
```diff
@@ -124,6 +124,21 @@ def audit_urls
       end
     end
 
+    def audit_head_branch
+      return if !@online || !@strict || spec_name != :head || !Utils::Git.remote_exists?(url)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/resource_auditor.rb:128]: Can this get split into separate `return`s? It's a little hard to follow as-is.

File: Library/Homebrew/resource_auditor.rb [hunk 2]
```diff
@@ -124,6 +124,21 @@ def audit_urls
...
+
+      branch = Utils.popen_read("git", "ls-remote", "--symref", url, "HEAD")
+                    .match(%r{ref: refs/heads/(.*?)\s+HEAD})[1]
+
+      return if branch == "master" && specs[:branch].blank? || branch == specs[:branch]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/resource_auditor.rb:133]: Can this be split into two `return`s or use `()` to make precedence clearer.

File: Library/Homebrew/resource_auditor.rb [hunk 3]
```diff
@@ -124,6 +124,21 @@ def audit_urls
...
+
+      if branch == "master"
+        problem "Remove `branch: \"#{specs[:branch]}\"`"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/resource_auditor.rb:136]: I'm tempted to say we want the branch to be always set so that it'll make it easier for us to change the default in future. Thoughts?

**Comment by nandahkrishna** [comment]
I realised that for some formulae we want the specified branch to not be the default branch, for example, for versioned formulae where different versions could have different development branches. I don't think this would be the case for too many formulae, so I think I'll add an allowlist for the audit, along with the suggested changes.

File: Library/Homebrew/resource_auditor.rb (hunk 3 above)
Comment (Rylan12) [reply Library/Homebrew/resource_auditor.rb:136]: Big 👍 to this

**Comment by Rylan12** [comment]
> I realised that for some formulae we want the specified branch to not be the default branch – like versioned formulae where different versions have different development branches. I don't think this would be the case for too many formulae, so I think I'll add an allowlist for the audit, along with the suggested changes.

Yeah, this is definitely a good idea to have. I'd say go ahead and continue with this route and if it turns out there are a ton of formulae on the list, we can change the audit to just be "you need to specify a branch name"

**Comment by Rylan12** [comment]
Could this be done for `stable`, too?

Also, I'm tempted to not make this a normal audit over `--strict`. Would that be unreasonable because of the number of formulae that would need to be changed? What about making the audit is `--strict` if the default branch is `master` and not `--strict` otherwise (and then we can plan on removing the `--strict` in the future once more formulae are updated.

**Comment by nandahkrishna** [comment]
> Could this be done for `stable`, too?

Stable URLs always specify a revision and mostly have a tag, so I don't know if this would be necessary. Hope there isn't some case I'm forgetting about here.

> Would that be unreasonable because of the number of formulae that would need to be changed?

I haven't yet checked the number of formulae that need to be updated, but I thought keeping this as a strict audit would be better because:
* Failing CI for an effectively syntax-only change seems undesirable. Those who use `HEAD`-installs and other interested contributors could volunteer to make the change in a separate syntax-only PR anyway.
* It will still be checked for `brew audit --new`, so new formulae have the correct default branch specified.

Seems like specifying the branch anyway (whether it's `master` or not) is better in the long run. I'll tweak this so that a branch must be specified, and add an allowlist which we would use when the default branch is intentionally not used for the formula (versioned formula with `head` and other cases).

**Comment by nandahkrishna** [comment]
Apologies for the late update, the latest commit makes specifying a branch mandatory for `head` and also adds an allowlist for specifying formulae where we intentionally use non-default branches. I tested this with some formulae:
* `sdl_mixer` for an intentional non-default branch.
* `zsh-vi-mode` to ensure that specifying a branch is mandatory.
* Modifying the head branch to a formula and ensuring the correct default branch is reported by the audit.

The allowlist, located at `audit_exceptions/head_non_default_branch_allowlist.json` in a Tap would look like this:
```json
{
    "sdl_mixer": "SDL-1.2",
    ...
}
```
Only one issue remains unsolved – `head` resources that specify non-default branches intentionally. I'll need to look for examples but here's a hypothetical one:
```ruby
head do
  ...

  resource "example-resource" do
    url "https://github.com/example-resource/example-resource.git", branch: "example"
  end
end
```
At the moment, the audit would raise an error if `example` isn't the default branch. We'll have to extend the allowlist to account for such cases as well, but I'm not sure what a good format/schema would be, looking forward to your suggestions. Here's one potential format which looks complicated, but such formulae should be very rare:
```json
{
    "sdl_mixer": "SDL-1.2",
    "example-formula": {
        "branch": "dev",
        "resources": {
            "example-resource": "example"
        }
    }
}
```

File: Library/Homebrew/resource_auditor.rb [hunk 4]
```diff
@@ -124,6 +124,20 @@ def audit_urls
       end
     end
 
+    def audit_head_branch
+      return unless @online
+      return unless @strict
+      return unless spec_name == :head
```
Comment (MikeMcQuaid) [inline Library/Homebrew/resource_auditor.rb:130]:
```suggestion
      return if spec_name != :head
```

**Review by MikeMcQuaid — APPROVED** [review]
Nice work! One tiny suggested tweak.

**Review by Rylan12 — APPROVED** [review]
> but such formulae should be very rare

I think it's fine to just figure that out if we encounter it, then.

Thanks, @nandahkrishna!

**Review by MikeMcQuaid — APPROVED** [review]

**Comment by vladimyr** [comment]
@nandahkrishna Just for the record I still feel bad sending you down the rabbit hole. I really didn't mean to and I honestly wish I was able to help you.
Here goes a **big** thank you to @nandahkrishna and the rest of the team! Great job, as usual! ❤️

**Comment by nandahkrishna** [comment]
No problem at all @vladimyr, happy to have worked on this 😄.
